### PR TITLE
feat: add `gettxout` and `getmempoolentry` endpoints + tests

### PR DESCRIPTION
**Pull request by cylewitruk**
## Description

Related to (prerequisite) #759 

## Changes

Adds implementation and tests for Bitcoin-core's `gettxout` and `getmempoolentry` RPC endpoints.

While `gettxout` proved to be unreliable when trying to calculate fees for an unknown potential mixture of confirmed/unconfirmed transactions, if you know whether or not its confirmed or not then it works fine (see tests). So it's not going to be used in #759 but I decided to leave it in as it can be useful.

Reverts bitcoin core to 25.1 in docker compose files.

## Testing Information

Added a number of tests testing both methods.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/bitcoin/rpc.rs [hunk 1]
```diff
@@ -389,9 +404,11 @@ impl BitcoinCoreClient {
     ///
     /// # Notes
     ///
-    /// - This method requires bitcoin-core v27 or later.
+    /// - This method requires bitcoin-core v25 or later.
     /// - The RPC endpoint does not in itself return raw transaction data, so
     ///   [`Self::get_tx`] must be used to fetch each transaction separately.
+    /// - Implementation based on documentation at
+    ///   https://bitcoincore.org/en/doc/27.0.0/rpc/blockchain/getmempooldescendants/
     pub fn get_mempool_descendants(&self, txid: &Txid) -> Result<Vec<Txid>, Error> {
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/bitcoin/rpc.rs:412]: Wish the docs had like a "from version XX" or something.

[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/bitcoin/rpc.rs]: [9dbd16f](https://github.com/stacks-network/sbtc/pull/858/commits/9dbd16f7ad5f94d6d1ca056744ec3bf87f67febf)

File: signer/src/bitcoin/rpc.rs [hunk 2]
```diff
@@ -347,7 +360,9 @@ impl BitcoinCoreClient {
     ///
     /// # Notes
     ///
-    /// This method requires bitcoin-core v27 or later.
+    /// This method requires bitcoin-core v25 or later and is based on the
+    /// documentation at
+    /// https://bitcoincore.org/en/doc/27.0.0/rpc/blockchain/gettxspendingprevout/
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/bitcoin/rpc.rs]: [9dbd16f](https://github.com/stacks-network/sbtc/pull/858/commits/9dbd16f7ad5f94d6d1ca056744ec3bf87f67febf)

File: signer/src/bitcoin/mod.rs [hunk 3]
```diff
@@ -82,4 +85,23 @@ pub trait BitcoinInteract: Sync + Send {
...
+    fn get_transaction_output(
+        &self,
+        outpoint: &bitcoin::OutPoint,
+        include_mempool: bool,
+    ) -> impl Future<Output = Result<Option<GetTxOutResult>, Error>> + Send;
+
+    /// Gets the associated fees for the given transaction. It is expected that
+    /// the provided transaction is known to the Bitcoin core node, either
+    /// confirmed or in the mempool, otherwise an error will be returned.
+    fn get_transaction_fee(&self, tx: &Txid) -> impl Future<Output = Result<Fees, Error>> + Send;
```
Comment (cylewitruk) [inline signer/src/bitcoin/mod.rs:100]: Note that it looks like the return signature here looks like it's going to change as I'll also need vsize

File: signer/src/bitcoin/rpc.rs [hunk 4]
```diff
@@ -485,4 +532,46 @@ impl BitcoinInteract for BitcoinCoreClient {
...
+            .get_tx(txid)?
+            .ok_or(Error::BitcoinTxMissing(*txid, None))?;
+
+        let vsize = tx_info.tx.vsize();
+
+        // If the transaction is confirmed then the node has block-undo data available,
+        // and we can simply get the fee via `get_tx_info()`.
+        if tx_info.confirmations.is_some() {
+            let confirmed_tx_info = self
+                .get_tx_info(txid, &tx_info.block_hash.unwrap())?
```
Comment (cylewitruk) [inline signer/src/bitcoin/rpc.rs:555]: Whoops! `unwrap()` snuck in here -- this is fixed in part 3 (next PR).

**Comment by cylewitruk** [comment]
Just pulled in some changes i had made in my part 3 branch since they change a bit of this stuff.

File: signer/src/bitcoin/rpc.rs [hunk 5]
```diff
@@ -341,13 +368,43 @@ impl BitcoinCoreClient {
...
+            serde_json::to_value(txid).map_err(Error::JsonSerialize)?,
+            // This is the verbosity level. The acceptable values are 0, 1,
+            // and 2, and we want the 2 because it will include all the
+            // required fields of the type.
+            serde_json::Value::Number(serde_json::value::Number::from(2u32)),
+        ];
+
+        match self
+            .inner
+            .call::<BitcoinTxFeeInfo>("getrawtransaction", &args)
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/bitcoin/rpc.rs:386]: I actually don't have a preference, will resolve.

File: signer/src/bitcoin/rpc.rs [hunk 6]
```diff
@@ -485,4 +532,46 @@ impl BitcoinInteract for BitcoinCoreClient {
...
+
+    async fn get_transaction_fee(&self, txid: &bitcoin::Txid) -> Result<Fees, Error> {
+        let tx_info = self
+            .get_tx(txid)?
+            .ok_or(Error::BitcoinTxMissing(*txid, None))?;
+
+        let vsize = tx_info.tx.vsize();
+
+        // If the transaction is confirmed then the node has block-undo data available,
+        // and we can simply get the fee via `get_tx_info()`.
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/bitcoin/rpc.rs]: Yeah I'm not 100% on it either; there's more tests coming in part 3 so I hope that we can gain a little more confidence at least for our usage.

File: signer/src/bitcoin/rpc.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/bitcoin/rpc.rs:386]: I'll look at it in part 3 :+1:

**Pull request merged**